File: core/ct_scraper.py
from playwright.sync_api import sync_playwright
import pandas as pd

def fetch_ct_offers_df(monthly_usage=2000):
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False, slow_mo=200)
        page = browser.new_page()

        # 1. Load homepage (forces Cloudflare + session init)
        page.goto("https://www.energizect.com", wait_until="domcontentloaded")
        page.wait_for_timeout(1500)

        # 2. Navigate to Rate Board
        page.click("text=Rate Board")
        page.wait_for_timeout(1500)

        # 3. Navigate to Compare Rates
        page.click("text=Compare Rates")
        page.wait_for_timeout(2000)

        # 4. Now the REAL form is loaded — select dropdowns
        page.select_option("select[name='utility']", "1206")          # Eversource
        page.select_option("select[name='customerClass']", "1206")    # Business
        page.select_option("select[name='monthlyUsage']", str(monthly_usage))

        # 5. Click Search
        page.click("text=Search & Compare Rates")

        # 6. Wait for offer cards
        page.wait_for_selector(".list-item--ect-card", timeout=20000)

        # 7. Call the API from inside the browser
        api_url = (
            "https://www.energizect.com/ectr_search_api/offers"
            f"?customerClass[]=1206&monthlyUsage={monthly_usage}&planTypeEdc[]=1191"
        )

        data = page.evaluate(
            """async (api_url) => {
                const resp = await fetch(api_url, {
                    method: "GET",
                    headers: { "X-Requested-With": "XMLHttpRequest" }
                });
                return await resp.json();
            }""",
            api_url,
        )

        browser.close()

    offers = data.get("data", [])
    if not offers:
        raise RuntimeError("No offers returned from EnergizeCT API.")

    df = pd.DataFrame(offers)

    df = df.rename(
        columns={
            "supplierName": "supplier",
            "rate": "rate_cents_per_kwh",
            "term": "term_description",
            "renewableEnergy": "recs_pct",
            "monthlyCost": "monthly_cost_at_usage",
            "phone": "phone",
        }
    )

    return df

File: node_scraper/ct_scraper.js
const { chromium } = require("playwright");
const fs = require("fs");

async function debugCT() {
    const browser = await chromium.launch({
        headless: false,   // MUST be false
        slowMo: 150,       // Human-like interaction
    });

    const context = await browser.newContext({
        userAgent:
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
        viewport: { width: 1400, height: 900 }
    });

    const page = await context.newPage();

    await page.goto(
        "https://www.energizect.com/rate-board/compare-energy-supplier-rates",
        { waitUntil: "domcontentloaded" }
    );

    await page.waitForTimeout(3000);

    // Scroll to modal buttons
    await page.evaluate(() => window.scrollBy(0, 400));

    // Click Eversource Business using real mouse events
    await page.locator("text=Eversource Business").click({ force: true });

    await page.waitForTimeout(2000);

    // Click Search & Compare Rates
    await page.locator("text=Search & Compare Rates").click({ force: true });

    // Wait for AJAX-loaded offer cards
    await page.waitForSelector(".list-item--ect-card", { timeout: 20000 });

    // Extract offer section
    const offerHTML = await page.evaluate(() => {
        const container =
            document.querySelector(".rate-board-results") ||
            document.querySelector(".view-content") ||
            document.body;

        return container.innerHTML;
    });

    fs.writeFileSync("ct_debug_section.html", offerHTML);

    console.log("WROTE ct_debug_section.html — send me this file.");
}

debugCT();